(fix) colima: fix return code handling (followup to #3097)

tobitege opened the pull request:
**What is the problem that this fixes or functionality that this introduces? Does it fix any open issues?**

Fixes #3097 

- This fixes the exit code of the `colima start` command being passed to the outer loop.
- Added a 10 second delay before retry.
- Increased to 5 total attempts (4 retries).
- Added "--verbose" flag and lowered random port range to 10000-20000

tobitege commented:
@SmartManoj This should work better now, right?

SmartManoj commented on .github/workflows/run-unit-tests.yml:
```diff
@@ -89,17 +89,20 @@ jobs:
 
-          # Attempt to start Colima
-          ATTEMPT_LIMIT=3
+          # Attempt to start Colima; 4 retries:
```
```suggestion
          # Attempt to start Colima; 5 retries:
```

tobitege replied:
Read the code: 5 total attempts, which means 4 retries.

SmartManoj commented on .github/workflows/run-unit-tests.yml:
```diff
@@ -89,27 +89,32 @@ jobs:
             else
               echo "Failed to start Colima. Attempt $i/$ATTEMPT_LIMIT."
               if [ $i -eq $ATTEMPT_LIMIT ]; then
                 colima delete
```
It should be executed before the last attempt.

tobitege commented on .github/workflows/run-unit-tests.yml:
```diff
@@ -89,27 +89,32 @@ jobs:
               break
             else
               echo "Failed to start Colima. Attempt $i/$ATTEMPT_LIMIT."
               if [ $i -eq $ATTEMPT_LIMIT ]; then
                 colima delete
```
Why? The loop goes `i<=ATTEMPT_LIMIT`, so it should delete when last attempt failed.

SmartManoj replied:
Deleting colima after the last attempt is no use.

tobitege replied:
I think I can actually remove the delete completely, it is rather pointless now.

SmartManoj replied:
It is a fallback method. https://github.com/OpenDevin/OpenDevin/pull/3061#issuecomment-2244160764

SmartManoj replied:
> Meaning: I doubt that a delete between failed attempts has worked before.

delete between failed attempts itself has not happened before.

tobitege replied:
I mean, we wouldn't _see_ that if all went well, but so far if colima failed, the delete didn't fix it 'cause then QEMU would've already terminated and what else happens then, unrecoverable state, I think.

SmartManoj replied:
> delete between failed attempts itself has not happened before. [#3061 (files)](https://github.com/OpenDevin/OpenDevin/pull/3061/files#r1691249029)

colima is not started after delete

tobitege replied:
So delete and re-install everytime? 
This isn't a fallback, that's just a hack.

SmartManoj replied:
> So delete and re-install everytime?

No. deleting only for the last attempt

SmartManoj commented on .github/workflows/run-unit-tests.yml:
```diff
@@ -87,35 +87,53 @@ jobs:
+              if [ $i -lt $ATTEMPT_LIMIT ]; then
+                echo "Stopping Colima forcefully..."
                 colima stop -f
+                ecthe "ho "Waiting 10 seconds before next attempt..."
```
```suggestion
                echo "Waiting 10 seconds before next attempt..."
```

tobitege replied:
😂  yeh, just saw that

tobitege commented:
@SmartManoj I've removed the "docker context" approach again.
Trying the loop with install and delete now. Also added that the IP will increase with every attempt.
From what I've seen thus far, it is the local IP address that it can't "resolve" or is blocked from using.
Bit of trial and error... 😔

tobitege commented:
Oops a copy/paste error, one sec....

tobitege commented:
> Github runner has only one IP. right?

I don't know, will see where it fails 😂

SmartManoj commented:
> Github runner has only one IP. right?

For a single runner during execution. [Docs](https://docs.github.com/en/actions/using-github-hosted-runners/about-github-hosted-runners/about-github-hosted-runners#ip-addresses)

tobitege commented:
> > Github runner has only one IP. right?
> 
> For a single runner during running. [Docs](https://docs.github.com/en/actions/using-github-hosted-runners/about-github-hosted-runners/about-github-hosted-runners#ip-addresses)

Thanks. I'll give it one more try with `limactl create colima-${IP_ADDRESS}` next and if that fails, will revert that IP changing as well.

tobitege commented:
> > Github runner has only one IP. right?
> 
> For a single runner during running. [Docs](https://docs.github.com/en/actions/using-github-hosted-runners/about-github-hosted-runners/about-github-hosted-runners#ip-addresses)

But here we are using a local IP address only, not an external IP. So the docs don't apply in this case, I think.

SmartManoj commented on .github/workflows/run-unit-tests.yml:
```diff
@@ -75,47 +75,56 @@ jobs:
+            # Create Lima instance with the specific IP address
+            limactl create colima-${IP_ADDRESS}
+
+            if ! colima start --network-address $IP_ADDRESS --arch x86_64 --cpu=1 --memory=1 --verbose --ssh-port $RANDOM_PORT; then
```
`network-address` is a boolean arg
https://github.com/abiosoft/colima/blob/c8c4c5a69b4e422dab76ac0a0f81515094302c2e/cmd/start.go#L160

tobitege replied:
Again, you have to read the full source:
https://github.com/abiosoft/colima/blob/c8c4c5a69b4e422dab76ac0a0f81515094302c2e/cmd/start.go#L159-L161
```py
	if util.MacOS() {
		startCmd.Flags().BoolVar(&startCmdArgs.Network.Address, "network-address", false, "assign reachable IP address to the VM")
	}
```

SmartManoj replied:
This line sets up a boolean command-line flag named network-address. When this flag is set to true by the user, it indicates that a reachable IP address should be assigned to the VM. By default, this flag is false. isn't it?

tobitege replied:
🤕 ohh, yes, I did just read the text, not the "BoolVar" and misunderstood it. 
There is not enough ☕☕☕around, apparently.
Thanks for correcting me there!

tobitege commented:
oh well, doesn't work with limactl neither, so going to remove that again ...

xingyaoww reviewed (approved):
Let's give it a try!

WEI567 commented:
可望打造新的
